wcpredictor/src/tournament.py

```diff
@@ -426,69 +437,230 @@ def __str__(self) -> str:
+        self.bracket = self._init_bracket(year)
+        self.verbose = verbose
+
+    def _parse_resume_from(self, resume_from, year):
+        if resume_from is None:
+            # starting from the beginning, i.e. the group stage
+            return pd.to_datetime(f"{year}-1-1"), "Group"
+        if resume_from in STAGES:
+            # end date from tournament round fixture dates
+            dates = pd.to_datetime(self.fixtures_df["date"])
+            # round start date
+            resume_date = dates[self.fixtures_df["stage"] == resume_from].min()
+            return resume_date, resume_from
+        if resume_from == "latest":
+            resume_from = pd.to_datetime(date.today())
+        dates = pd.to_datetime(self.fixtures_df["date"])
+        if sum(dates >= resume_from) == 0:
+            raise ValueError(
+                f"There are no more fixtures after resume_from = {resume_from}"
+            )
+        resume_stage = self.fixtures_df[dates >= resume_from].iloc[0]["stage"]
+        return resume_from, resume_stage
+
+    def _fill_played_fixtures(self, year):
+        """Fill the actual results of played results up to resume_from, and return
+        the stage that takes us up to"""
+        self.fixtures_df["actual_home"] = np.nan
+        self.fixtures_df["actual_away"] = np.nan
+
+        end_date = pd.to_datetime(self.resume_date) - pd.Timedelta(days=1)
+        actual_results, _ = get_results_data(
+            start_date=f"{year}-01-01",
+            end_date=end_date,
+            competitions="W",
+        )
+
+        for _, result in actual_results.iterrows():
+            fixture_idx = np.where(
+                (self.fixtures_df["home_team"] == result["home_team"])
+                & (self.fixtures_df["away_team"] == result["away_team"])
+                & (self.fixtures_df["date"] == result["date"])
+            )[0]
+            if len(fixture_idx) == 0:
+                continue
+            if len(fixture_idx) > 1:
+                raise RuntimeError(
+                    f"Found multiple {result['home_team']} vs. "
+                    f"{result['away_team']} fixtures on {result['date']}"
+                )
+            self.fixtures_df.loc[
+                self.fixtures_df.index[fixture_idx[0]], "actual_home"
+            ] = result["home_score"]
+            self.fixtures_df.loc[
+                self.fixtures_df.index[fixture_idx[0]], "actual_away"
+            ] = result["away_score"]
+
+    def _init_bracket(self, year):
+        aliases = get_alias_data(year)
+        # not na actual score and not an actual team name
+        bracket = pd.DataFrame(
+            index=np.arange(self.num_samples), columns=aliases["team"].index.values
+        )
+        if self.resume_stage == "Group":
+            return bracket
+
+        for aka, team in aliases["team"].items():
+            # only fill up to resume_date and resume_stage
+            if (self.resume_stage == "R16") and (len(aka) > 2):
```

Comment:
I think this logic means that we can't resume from partway through e.g. the round of 16 - even if there are some aliases filled in, we won't use them.
But maybe that's fine - it's just a few days waiting...